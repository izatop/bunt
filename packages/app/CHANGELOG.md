# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.7.0](https://github.com/izatop/typesafeunit/compare/v0.6.28...v0.7.0) (2020-05-26)


### Features

* new eslint rules ([62417eb](https://github.com/izatop/typesafeunit/commit/62417eb1385862c93108991075f864536b90a748))





## [0.6.28](https://github.com/izatop/typesafeunit/compare/v0.6.27...v0.6.28) (2020-05-17)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.27](https://github.com/izatop/typesafeunit/compare/v0.6.26...v0.6.27) (2020-05-15)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.26](https://github.com/izatop/typesafeunit/compare/v0.6.25...v0.6.26) (2020-05-12)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.22](https://github.com/izatop/typesafeunit/compare/v0.6.21...v0.6.22) (2020-05-07)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.21](https://github.com/izatop/typesafeunit/compare/v0.6.20...v0.6.21) (2020-05-06)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.20](https://github.com/izatop/typesafeunit/compare/v0.6.19...v0.6.20) (2020-05-06)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.19](https://github.com/izatop/typesafeunit/compare/v0.6.18...v0.6.19) (2020-05-01)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.18](https://github.com/izatop/typesafeunit/compare/v0.6.17...v0.6.18) (2020-04-30)


### Bug Fixes

* logging ([2365a0b](https://github.com/izatop/typesafeunit/commit/2365a0b12902300b2e898e2c07aa5769be3b8e56))





## [0.6.15](https://github.com/izatop/typesafeunit/compare/v0.6.14...v0.6.15) (2020-04-04)


### Bug Fixes

* action state ([32974d7](https://github.com/izatop/typesafeunit/commit/32974d7ae0723888531d376d2642d9d62e0954f8))





## [0.6.14](https://github.com/izatop/typesafeunit/compare/v0.6.13...v0.6.14) (2020-04-03)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.13](https://github.com/izatop/typesafeunit/compare/v0.6.12...v0.6.13) (2020-04-01)


### Bug Fixes

* runtime, tests ([c17980f](https://github.com/izatop/typesafeunit/commit/c17980fb75e047dfc001e6fca917ef1f3a9d84dd))





## [0.6.12](https://github.com/izatop/typesafeunit/compare/v0.6.11...v0.6.12) (2020-04-01)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.11](https://github.com/izatop/typesafeunit/compare/v0.6.10...v0.6.11) (2020-04-01)


### Bug Fixes

* action return type ([fe2dc96](https://github.com/izatop/typesafeunit/commit/fe2dc964e3e0776f15288f511cf7962cbf3fa8a3))





## [0.6.10](https://github.com/izatop/typesafeunit/compare/v0.6.9...v0.6.10) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.9](https://github.com/izatop/typesafeunit/compare/v0.6.8...v0.6.9) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.8](https://github.com/izatop/typesafeunit/compare/v0.6.7...v0.6.8) (2020-03-30)


### Bug Fixes

* errors ([5c61c09](https://github.com/izatop/typesafeunit/commit/5c61c09f19ddb0415238d8cd48681b9dddd74f29))





## [0.6.7](https://github.com/izatop/typesafeunit/compare/v0.6.6...v0.6.7) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.6](https://github.com/izatop/typesafeunit/compare/v0.6.5...v0.6.6) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.5](https://github.com/izatop/typesafeunit/compare/v0.6.4...v0.6.5) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.4](https://github.com/izatop/typesafeunit/compare/v0.6.3...v0.6.4) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.3](https://github.com/izatop/typesafeunit/compare/v0.6.2...v0.6.3) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.2](https://github.com/izatop/typesafeunit/compare/v0.6.1...v0.6.2) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.1](https://github.com/izatop/typesafeunit/compare/v0.6.0...v0.6.1) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





# [0.6.0](https://github.com/izatop/typesafeunit/compare/v0.5.2...v0.6.0) (2020-03-30)


### Features

* logger, perf etc ([daed7fe](https://github.com/izatop/typesafeunit/commit/daed7fe86ea6e9cb3bd90830a77167e6f9e21755))





## [0.5.2](https://github.com/izatop/typesafeunit/compare/v0.5.1...v0.5.2) (2020-03-26)

**Note:** Version bump only for package @typesafeunit/app





## [0.5.1](https://github.com/izatop/typesafeunit/compare/v0.5.0...v0.5.1) (2020-03-26)

**Note:** Version bump only for package @typesafeunit/app





# [0.5.0](https://github.com/izatop/typesafeunit/compare/v0.4.6...v0.5.0) (2020-03-26)

**Note:** Version bump only for package @typesafeunit/app





## [0.4.5](https://github.com/izatop/typesafeunit/compare/v0.4.4...v0.4.5) (2020-03-25)


### Bug Fixes

* improve application factory ([0899aa1](https://github.com/izatop/typesafeunit/commit/0899aa1329fb189812af65af6580cc228145d75c))





## [0.4.3](https://github.com/izatop/typesafeunit/compare/v0.4.2...v0.4.3) (2020-03-25)

**Note:** Version bump only for package @typesafeunit/app





## [0.4.2](https://github.com/izatop/typesafeunit/compare/v0.4.1...v0.4.2) (2020-03-25)

**Note:** Version bump only for package @typesafeunit/app





## [0.4.1](https://github.com/izatop/typesafeunit/compare/v0.4.0...v0.4.1) (2020-03-25)

**Note:** Version bump only for package @typesafeunit/app





# [0.4.0](https://github.com/izatop/typesafeunit/compare/v0.3.9...v0.4.0) (2020-03-25)


### Features

* http server ([62abc0f](https://github.com/izatop/typesafeunit/commit/62abc0f20cfefac0a7d5fa6bc91c030719d5e9a8))





## [0.3.9](https://github.com/izatop/typesafeunit/compare/v0.3.8...v0.3.9) (2020-03-08)

**Note:** Version bump only for package @typesafeunit/app





## [0.3.8](https://github.com/izatop/typesafeunit/compare/v0.3.7...v0.3.8) (2020-03-08)

**Note:** Version bump only for package @typesafeunit/app





## [0.3.7](https://github.com/izatop/typesafeunit/compare/v0.3.6...v0.3.7) (2020-03-08)

**Note:** Version bump only for package @typesafeunit/app





## [0.3.6](https://github.com/izatop/typesafeunit/compare/v0.3.5...v0.3.6) (2020-03-08)


### Bug Fixes

* error handling ([1792e2b](https://github.com/izatop/typesafeunit/commit/1792e2b83fec230a396d955ab12ccea96e7bc758))





## [0.3.5](https://github.com/izatop/typesafeunit/compare/v0.3.4...v0.3.5) (2020-03-07)


### Bug Fixes

* headers keys case sensitivity ([37c372d](https://github.com/izatop/typesafeunit/commit/37c372d23a1ac9a5175aa6428c3b1c82e17e669c))
* **headers:** case sensitivity fix ([c119eae](https://github.com/izatop/typesafeunit/commit/c119eae6fb96fba11f27c260f130b351addc34fa))





## [0.3.4](https://github.com/izatop/typesafeunit/compare/v0.3.3...v0.3.4) (2020-03-07)


### Bug Fixes

* type fix ([a41f766](https://github.com/izatop/typesafeunit/commit/a41f76606330f83a8355081fdcc7d9c112708efe))





## [0.3.3](https://github.com/izatop/typesafeunit/compare/v0.3.2...v0.3.3) (2020-03-05)


### Bug Fixes

* **npm:** add LICENSE ([d7bf57f](https://github.com/izatop/typesafeunit/commit/d7bf57fe2dcb9f83cc4a570c99fafcb5e318bdd4))





## [0.3.2](https://github.com/izatop/typesafeunit/compare/v0.3.1...v0.3.2) (2020-03-05)

**Note:** Version bump only for package @typesafeunit/app





## [0.3.1](https://github.com/izatop/typesafeunit/compare/v0.3.0...v0.3.1) (2020-03-05)

**Note:** Version bump only for package @typesafeunit/app





# 0.3.0 (2020-03-05)


### Features

* initial commit ([a76a484](https://github.com/izatop/typesafeunit/commit/a76a4849282c382311f4d0cb9f9bd66a10bcceaa))
