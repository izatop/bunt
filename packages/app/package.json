{
    "name": "@typesafeunit/app",
    "version": "0.6.12",
    "keywords": [
        "typescript"
    ],
    "author": {
        "name": "Artur Bier",
        "email": "izatop@gmail.com"
    },
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist/"
    ],
    "description": "Application Server",
    "repository": "git@github.com:izatop/typesafeunit.git",
    "scripts": {
        "clean": "yarn build:clean",
        "watch": "yarn build:watch",
        "build": "yarn build:clean && tsc",
        "build:clean": "rimraf dist tsconfig.tsbuildinfo",
        "build:watch": "yarn build --watch"
    },
    "publishConfig": {
        "access": "public"
    },
    "dependencies": {
        "@typesafeunit/unit": "^0.6.12",
        "@typesafeunit/util": "^0.6.12",
        "path-to-regexp": "^6.1.0"
    },
    "license": "MIT"
}
