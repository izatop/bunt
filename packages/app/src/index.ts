export * from "./Payload/index.js";
export * from "./Route/index.js";
export * from "./Request/index.js";
export * from "./TransformRequest/index.js";
export * from "./Application.js";
export * from "./interfaces.js";
