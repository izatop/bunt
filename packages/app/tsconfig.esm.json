{
  "extends": "../../tsconfig.lib.json",
  "compilerOptions": {
    "module": "ESNext",
    "moduleResolution": "Node",
    "rootDir": "src",
    "outDir": "dist/esm"
  },
  "references": [
    {
      "path": "../type/tsconfig.esm.json"
    },
    {
      "path": "../assert/tsconfig.esm.json"
    },
    {
      "path": "../async/tsconfig.esm.json"
    },
    {
      "path": "../input/tsconfig.esm.json"
    },
    {
      "path": "../is/tsconfig.esm.json"
    },
    {
      "path": "../unit/tsconfig.esm.json"
    },
    {
      "path": "../util/tsconfig.esm.json"
    }
  ],
  "include": [
    "src"
  ]
}
