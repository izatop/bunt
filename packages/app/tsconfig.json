{
    "extends": "../../tsconfig.lib.json",
    "compilerOptions": {
        "rootDir": "src",
        "outDir": "dist"
    },
    "references": [
        {
            "path": "../util"
        },
        {
            "path": "../unit"
        },
        {
            "path": "../input"
        }
    ],
    "include": [
        "src"
    ]
}
