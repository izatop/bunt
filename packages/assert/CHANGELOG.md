# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.28.1](https://github.com/izatop/bunt/compare/v0.28.0...v0.28.1) (2023-06-28)


### Bug Fixes

* update deps, fix imports (.js) ([00a1d1c](https://github.com/izatop/bunt/commit/00a1d1c2ab0ed2ed6c98ea9387a4702682f7cbab))





# [0.28.0](https://github.com/izatop/bunt/compare/v0.27.1...v0.28.0) (2023-06-22)

**Note:** Version bump only for package @bunt/assert





## [0.27.1](https://github.com/izatop/bunt/compare/v0.27.0...v0.27.1) (2023-06-15)

**Note:** Version bump only for package @bunt/assert





# [0.27.0](https://github.com/izatop/bunt/compare/v0.26.3...v0.27.0) (2023-06-14)


### Features

* task runner ([b108fc0](https://github.com/izatop/bunt/commit/b108fc08afb75dce76dfd76fabe00f8af4d8ebe7))
