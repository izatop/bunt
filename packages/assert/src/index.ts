export * from "./Exception/index.js";
export * from "./assert.js";
