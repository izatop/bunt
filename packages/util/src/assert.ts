import {AssertionError} from "./Exception";
import {isFunction} from "./is";

export type AssertionDetailsAllowType = string | object | null | number;
export type AssertionDetails = (() => AssertionDetailsAllowType) | AssertionDetailsAllowType;
export type AssertionMessage = string | (() => string);

export function assert(expr: any, message?: AssertionMessage, details?: AssertionDetails): asserts expr {
    if (!expr) {
        throw new AssertionError(isFunction(message) ? message() : message, details);
    }
}

export function fails(expr: any, message?: AssertionMessage, details?: AssertionDetails) {
    if (!!expr) {
        throw new AssertionError(isFunction(message) ? message() : message, details);
    }
}

export function pass<T>(value: any, transform: (v: any) => T, validator: (v: any) => boolean, label?: string): T {
    assert(value, `${label || "Value"} should be defined`);

    const nextValue = transform(value);
    assert(
        validator(nextValue),
        `This ${label || "value"} "${nextValue}" should pass the validation stage: ${validator.toString()}`,
    );

    return nextValue;
}
