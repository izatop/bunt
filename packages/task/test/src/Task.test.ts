import {assert} from "@bunt/assert";
import {
    ConcurrencyLimiter,
    FixedWindowLimiter,
    ITaskDone,
    ITaskReject,
    ITaskResult,
    ITaskSchedule,
    TaskState,
} from "../../src";
import {TestContext} from "./app/TestContext";
import {TestPayload, TestQueue} from "./app/TestQueue";
import {TestRunner} from "./app/TestRunner";
import TestTask from "./app/TestTask";

const pending: {(): void}[] = [];
const add = (): Promise<void> => new Promise((res) => pending.push(res));

describe("Task", () => {
    test("Runner", async () => {
        const dataset: TestPayload[] = [
            [
                {state: 0, runAt: new Date()},
                {id: 1, value: 1},
            ],
        ];

        const queue = new TestQueue(dataset);
        const runner = await TestRunner.factory(new TestContext(queue), {polling: false});
        runner.register(TestTask);
        await runner.handle();

        expect(queue.find(1)?.[0].state).toBe(TaskState.DONE);
        expect(queue.disposed).toBe(true);
    });

    test("Runner.enqueue", async () => {
        const dataset: TestPayload[] = [
            [
                {state: 0, runAt: new Date()},
                {id: 1, value: 1},
            ],
        ];

        const queue = new TestQueue(dataset);
        const runner = await TestRunner.factory(new TestContext(queue), {polling: false});

        const iterator = await TestTask.factory(runner.context);
        const payload = await iterator.next();
        expect(payload.value).toBeDefined();

        assert(payload.value);
        await runner.enqueue(TestTask, payload.value);
        expect(queue.find(1)?.[0].state).toBe(TaskState.DONE);

        await expect(iterator.next()).resolves.toEqual({done: true});
        expect(queue.disposed).toBe(true);
    });

    test("Runner.getTaskUpgradeState", async () => {
        const queue = new TestQueue([]);
        const runner = await TestRunner.factory(new TestContext(queue));

        const result = "done";
        const date = new Date("2000-01-01T00:00:00.000Z");

        const schedule: ITaskSchedule = {
            date,
            result,
            state: TaskState.READY,
            reason: "ok",
        };

        const rejected: ITaskReject = {
            state: TaskState.REJECTED,
            reason: "Unexpected error",
        };

        const done: ITaskDone = {
            result,
            state: TaskState.DONE,
        };

        const simple: ITaskResult = {result, reason: "success"};

        expect(runner.getTaskUpgradeState(schedule)).toMatchSnapshot();
        expect(runner.getTaskUpgradeState(rejected)).toMatchSnapshot();
        expect(runner.getTaskUpgradeState(done)).toMatchSnapshot();
        expect(runner.getTaskUpgradeState(simple)).toMatchSnapshot();
        expect(runner.getTaskUpgradeState()).toMatchSnapshot();
    });

    test("ConcurrencyLimiter", async () => {
        const dataset = [[1, void 0], [2, Promise.resolve()]];
        const limiter = new ConcurrencyLimiter(2);
        for (const [value, ret] of dataset) {
            limiter.enqueue(add());
            expect(limiter.size).toBe(value);
            expect(limiter.limit()).toEqual(ret);
        }

        const limitedPromise = limiter.limit();
        pending.forEach((fn) => fn());

        await expect(limitedPromise).resolves.toBeUndefined();
    });

    test("FixedWindowLimiter", async () => {
        const dataset = [[1, void 0], [2, Promise.resolve()]];
        const limiter = new FixedWindowLimiter(10, 2, 10);
        for (const [value, ret] of dataset) {
            limiter.enqueue(add());
            expect(limiter.size).toBe(value);
            expect(limiter.limit()).toEqual(ret);
        }

        const limitedPromise = limiter.limit();
        pending.forEach((fn) => fn());

        await expect(limitedPromise).resolves.toBeUndefined();
    });
});
