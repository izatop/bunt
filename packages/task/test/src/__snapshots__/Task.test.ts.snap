// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Runner.getTaskUpgradeState 1`] = `
{
  "reason": "ok",
  "result": "done",
  "runAt": 2000-01-01T00:00:00.000Z,
  "state": 0,
}
`;

exports[`Task Runner.getTaskUpgradeState 2`] = `
{
  "reason": "Unexpected error",
  "state": 2,
}
`;

exports[`Task Runner.getTaskUpgradeState 3`] = `
{
  "result": "done",
  "state": 3,
}
`;

exports[`Task Runner.getTaskUpgradeState 4`] = `
{
  "reason": "success",
  "result": "done",
  "state": 3,
}
`;

exports[`Task Runner.getTaskUpgradeState 5`] = `
{
  "state": 3,
}
`;
